tests/sterm/test_terminal.py

```diff
@@ -0,0 +1,213 @@
+
+    terminal.run(port, baud)
+
+    mock_sterm.assert_called_once_with(mock_serial(), echo=True)
+
+
+def test_starts_sterm_thread(mock_sterm, mock_serial):
+    terminal.run("tty.122", 9600)
+
+    mock_sterm().start.assert_called_once()
+
+
+def test_joins_tx_and_rx_threads(mock_sterm, mock_serial):
+    terminal.run("tty.122", 9600)
+
+    mock_sterm().join.assert_any_call(True)
+
+
+def test_joins_tx_thread_after_keyboard_interrupt(mock_sterm, mock_serial):
+    mock_sterm().join.side_effect = (KeyboardInterrupt(), None)
+
+    terminal.run("tty.122", 9600)
+
+    mock_sterm().join.assert_called_with()
+
+
+def test_closes_sterm(mock_sterm, mock_serial):
+    terminal.run("tty.122", 9600)
+
+    mock_sterm().close.assert_called_once()
+
+
+def test_closes_sterm_after_exception(mock_sterm, mock_serial):
+    mock_sterm().join.side_effect = (Exception(), None)
+    with pytest.raises(Exception):
+        terminal.run("tty.122", 9600)
+
+    mock_sterm().close.assert_called_once()
+
+
+def test_sets_terminal_special_chars(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+
+    assert term.exit_character == terminal.CTRL_C
+    assert term.menu_character == terminal.CTRL_T
+    assert term.reset_character == terminal.CTRL_B
+    assert term.help_character == terminal.CTRL_H
+
+
+def test_sets_terminal_rx_and_tx_encoding_to_utf8(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+
+    assert term.input_encoding == "UTF-8"
+    assert term.output_encoding == "UTF-8"
+
+
+def test_stops_terminal_when_ctrl_c_received(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.alive = True
+    mock_console().getkey.return_value = terminal.CTRL_C
+
+    term.writer()
+
+    assert term.alive is False
+
+
+def test_stops_terminal_on_keyboard_interrupt(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.alive = True
+    mock_console().getkey.side_effect = KeyboardInterrupt()
+
+    term.writer()
+
+    assert term.alive is False
+
+
+@pytest.mark.parametrize("menu_key", terminal.VALID_MENU_KEYS)
+def test_handles_valid_menu_key(menu_key, mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.handle_menu_key = mock.Mock()
+    term.alive = True
+    mock_console().getkey.side_effect = (terminal.CTRL_T, menu_key, terminal.CTRL_C)
+
+    term.writer()
+
+    term.handle_menu_key.assert_called_once_with(menu_key)
+
+
+INVALID_MENU_KEYS = tuple(set(chr(i) for i in range(0, 127)) - set(terminal.VALID_MENU_KEYS) - set([terminal.CTRL_H]))
+
+
+@pytest.mark.parametrize("menu_key", INVALID_MENU_KEYS)
+def test_ignores_invalid_menu_key(menu_key, mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.handle_menu_key = mock.Mock()
+    term.alive = True
+    mock_console().getkey.side_effect = (terminal.CTRL_T, menu_key)
+
+    with pytest.raises(StopIteration):
```

Comment:
Why is a StopIteration an invalid menu key? Why not KeyboardError or something?

Reply:
`StopIteration` isn't a valid menu key, it's a way to get the terminal to stop blocking so I can assert on the state of the mock.